### PR TITLE
Docstrings for numpy ops

### PR DESCRIPTION
**Pull request by FayazRahman**
- Added docstrings for ops in `keras_core/ops/numpy.py`

### CODE REVIEW

File: keras_core/ops/numpy.py [hunk 1]
```diff
@@ -1446,6 +1526,27 @@ def compute_output_spec(self, x):
...
+        x: Input tensor.
+        axis: Axis or tuple of axes along which to count the number of
+            non-zeros. Defaults to None.
+
+    Returns:
+        int or tensor of ints.
+
+    Examples:
+    >>> x = keras_core.ops.array([[0, 1, 7, 0],
+    ...                [3, 0, 2, 19]])
```
Comment (fchollet) [inline keras_core/ops/numpy.py:1542]: This can likely fit on a single line.

File: keras_core/ops/numpy.py [hunk 2]
```diff
@@ -1500,6 +1601,38 @@ def compute_output_spec(self, x1, x2):
...
+
+    In cases where both input vectors have dimension 2, the z-component of
+    the cross product is returned.
+
+    Args:
+        x1: Components of the first vector(s).
+        x2: Components of the second vector(s).
+        axisa: Axis of `x1` that defines the vector(s). Defaults to -1.
+        axisb: Axis of `x2` that defines the vector(s). Defaults to -1.
+        axisc: Axis of `x` along which the cross product vector(s).
```
Comment (fchollet) [inline keras_core/ops/numpy.py:1623]: `x` isn't defined here

File: keras_core/ops/numpy.py [hunk 3]
```diff
@@ -1500,6 +1601,38 @@ def compute_output_spec(self, x1, x2):
...
+            Ignored if both input vectors have dimension 2, as the return is
+            scalar. By default, the last axis.
+        axis: If defined, the axis of a, b and c that defines the vector(s)
```
Comment (fchollet) [inline keras_core/ops/numpy.py:1626]: x1, x2

File: keras_core/ops/numpy.py [hunk 4]
```diff
@@ -1500,6 +1601,38 @@ def compute_output_spec(self, x1, x2):
 
 @keras_core_export(["keras_core.ops.cross", "keras_core.ops.numpy.cross"])
 def cross(x1, x2, axisa=-1, axisb=-1, axisc=-1, axis=None):
+    """Returns the cross product of two (arrays of) vectors.
+
+    The cross product of `x1` and `x2` in :math:`R^3` is a vector
```
Comment (fchollet) [inline keras_core/ops/numpy.py:1606]: Don't use `:math:`, that's not supported.

File: keras_core/ops/numpy.py [hunk 5]
```diff
@@ -1603,6 +1756,38 @@ def compute_output_spec(self, x):
 
 @keras_core_export(["keras_core.ops.diag", "keras_core.ops.numpy.diag"])
 def diag(x, k=0):
+    """Extract a diagonal or construct a diagonal array.
+
+    Args:
+        x: Input tensor. If `x` is 2-D, returns the k-th diagonal of `x`.
+            If `x` is 1-D, return a 2-D tensor with `x` on the k-th diagonal.
+        k: The diagonal to consider. Defaults to 0. Use k > 0 for diagonals
```
Comment (fchollet) [inline keras_core/ops/numpy.py:1764]: Use backticks around code keywords

File: keras_core/ops/numpy.py [hunk 6]
```diff
@@ -1953,6 +2297,15 @@ def compute_output_spec(self, x1, x2):
 
 @keras_core_export(["keras_core.ops.equal", "keras_core.ops.numpy.equal"])
 def equal(x1, x2):
+    """Returns (x1 == x2) element-wise.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2300]: Use backticks

File: keras_core/ops/numpy.py [hunk 7]
```diff
@@ -2245,6 +2698,18 @@ def compute_output_spec(self, n, dtype="float32"):
 
 @keras_core_export(["keras_core.ops.identity", "keras_core.ops.numpy.identity"])
 def identity(n, dtype="float32"):
+    """Return the identity tensor.
+
+    The identity tensor is a square tensor with ones on the main diagonal and
+    zeros elsewhere.
+
+    Args:
+        n: Number of rows (and columns) in the *n x n* output tensor.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2707]: Use backticks

File: keras_core/ops/numpy.py [hunk 8]
```diff
@@ -2276,6 +2749,16 @@ def compute_output_spec(self, x1, x2):
 
 @keras_core_export(["keras_core.ops.isclose", "keras_core.ops.numpy.isclose"])
 def isclose(x1, x2):
+    """Return a boolean tensor with elements representing whether two tensors
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2752]: The first line should fit on exactly one line

File: keras_core/ops/numpy.py [hunk 9]
```diff
@@ -2291,6 +2774,19 @@ def compute_output_spec(self, x):
 
 @keras_core_export(["keras_core.ops.isfinite", "keras_core.ops.numpy.isfinite"])
 def isfinite(x):
+    """Return a boolean tensor with elements representing whether each
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2777]: Same here

File: keras_core/ops/numpy.py [hunk 10]
```diff
@@ -2306,6 +2802,15 @@ def compute_output_spec(self, x):
 
 @keras_core_export(["keras_core.ops.isinf", "keras_core.ops.numpy.isinf"])
 def isinf(x):
+    """Test element-wise for positive or negative infinity and return result
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2805]: Same here

File: keras_core/ops/numpy.py [hunk 11]
```diff
@@ -2418,6 +2949,36 @@ def compute_output_spec(self, start, stop):
 def linspace(
     start, stop, num=50, endpoint=True, retstep=False, dtype=None, axis=0
 ):
+    """Return evenly spaced numbers over a specified interval.
+
+    Returns `num` evenly spaced samples, calculated over the interval
+    [`start`, `stop`].
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2955]: The entire interval should use backticks

File: keras_core/ops/numpy.py [hunk 12]
```diff
@@ -2471,6 +3048,17 @@ def compute_output_spec(self, x):
 
 @keras_core_export(["keras_core.ops.log1p", "keras_core.ops.numpy.log1p"])
 def log1p(x):
+    """Returns the natural logarithm of one plus the input tensor,
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3051]: Same here

File: keras_core/ops/numpy.py [hunk 13]
```diff
@@ -2529,6 +3137,16 @@ def compute_output_spec(self, x1, x2):
     ]
 )
 def logical_and(x1, x2):
+    """Computes the element-wise logical AND of the given input tensors.
+    Zeros are treated as False and non-zeros are treated as True.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3141]: Same here

File: keras_core/ops/numpy.py [hunk 14]
```diff
@@ -2572,6 +3199,16 @@ def compute_output_spec(self, x1, x2):
     ]
 )
 def logical_or(x1, x2):
+    """Computes the element-wise logical OR of the given input tensors.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3202]: Etc

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Please revert the changes related to running `black` on `examples/`, as they make the PR harder to review.

File: keras_core/ops/numpy.py [hunk 15]
```diff
@@ -1273,6 +1291,17 @@ def compute_output_spec(self, x):
 
 @keras_core_export(["keras_core.ops.ceil", "keras_core.ops.numpy.ceil"])
 def ceil(x):
+    """Return the ceiling of the input, element-wise.
+
+    The ceil of the scalar `x` is the smallest integer `i`, such that
+        `i >= x`.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:1297]: Unindent

File: keras_core/ops/numpy.py [hunk 16]
```diff
@@ -1446,6 +1528,27 @@ def compute_output_spec(self, x):
...
 )
 def count_nonzero(x, axis=None):
+    """Counts the number of non-zero values in `x` along the given `axis`.
+
+    If no axis is specified then all non-zeros in the tensor are counted.
+
+    Args:
+        x: Input tensor.
+        axis: Axis or tuple of axes along which to count the number of
+            non-zeros. Defaults to None.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:1538]: Backticks for None

File: keras_core/ops/numpy.py [hunk 17]
```diff
@@ -2418,6 +2949,36 @@ def compute_output_spec(self, start, stop):
...
+        start: The starting value of the sequence.
+        stop: The end value of the sequence, unless `endpoint` is set to
+            False. In that case, the sequence consists of all but the last
+            of `num + 1` evenly spaced samples, so that `stop` is excluded.
+            Note that the step size changes when `endpoint` is False.
+        num: Number of samples to generate. Default is 50. Must be
+            non-negative.
+        endpoint: If True, `stop` is the last sample. Otherwise, it is
+            not included. Default is True.
+        retstep: If True, return (`samples`, `step`), where `step` is the
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2969]: Here too

File: keras_core/ops/numpy.py [hunk 18]
```diff
@@ -2622,6 +3261,31 @@ def compute_output_spec(self, start, stop):
...
+    `base ** stop` (see `endpoint` below).
+
+    Args:
+        start: The starting value of the sequence.
+        stop: The final value of the sequence, unless `endpoint` is False.
+            In that case, `num + 1` values are spaced over the interval in
+            log-space, of which all but the last (a sequence of length `num`)
+            are returned.
+        num: Number of samples to generate. Default is 50.
+        endpoint: If True, `stop` is the last sample. Otherwise, it is not
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3276]: Here too

File: keras_core/ops/numpy.py [hunk 19]
```diff
@@ -2788,6 +3522,19 @@ def compute_output_spec(self, x):
...
 def min(x, axis=None, keepdims=False, initial=None):
+    """Return the minimum of a tensor or minimum along an axis.
+
+    Args:
+        x: Input tensor.
+        axis: Axis or axes along which to operate. By default, flattened input
+            is used.
+        keepdims: If this is set to True, the axes which are reduced are left
+            in the result as dimensions with size one. Default is False.
+        initial: The maximum value of an output element. Default is None.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3533]: Backticks for None

File: keras_core/ops/numpy.py [hunk 20]
```diff
@@ -3006,6 +3835,19 @@ def compute_output_spec(self, x1, x2):
 
 @keras_core_export(["keras_core.ops.outer", "keras_core.ops.numpy.outer"])
 def outer(x1, x2):
+    """Compute the outer product of two vectors.
+
+    Given two vectors `x1` and `x2`, the outer product is:
+
+        out[i, j] = x1[i] * x2[j]
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3842]: Use a bactick block here

File: keras_core/ops/numpy.py [hunk 21]
```diff
@@ -3060,6 +3902,35 @@ def compute_output_spec(self, x):
...
+            width for all axes.
+        mode: One of `"constant"`, `"edge"`, `"linear_ramp"`,
+            `"maximum"`, `"mean"`, `"median"`, `"minimum"`,
+            `"reflect"`, `"symmetric"`, `"wrap"`, `"empty"`,
+            `"circular"`. Default is `"constant"`.
+
+    Note:
+        Torch backend only supports modes `"constant"`, `"reflect"`,
+        `"symmetric"` and `"circular"`.
+
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3924]: Remove line break

File: keras_core/ops/numpy.py [hunk 22]
```diff
@@ -3092,6 +3963,20 @@ def compute_output_spec(self, x):
...
 @keras_core_export(["keras_core.ops.prod", "keras_core.ops.numpy.prod"])
 def prod(x, axis=None, keepdims=False, dtype=None):
+    """Return the product of tensor elements over a given axis.
+
+    Args:
+        x: Input tensor.
+        axis: Axis or axes along which a product is performed. The default,
+            `axis=None`, will compute the product of all elements
+            in the input tensor.
+        keepdims: If this is set to True, the axes which are reduce
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3973]: Backticks for True

File: keras_core/ops/numpy.py [hunk 23]
```diff
@@ -3715,6 +4850,20 @@ def compute_output_spec(self, N, M=None, k=0, dtype="float32"):
 
 @keras_core_export(["keras_core.ops.tri", "keras_core.ops.numpy.tri"])
 def tri(N, M=None, k=0, dtype="float32"):
+    """Return a tensor with ones at and below a diagonal and zeros elsewhere.
+
+    Args:
+        N: Number of rows in the tensor.
+        M: Number of columns in the tensor.
+        k: The sub-diagonal at and below which the array is filled.
+            k = 0 is the main diagonal, while k < 0 is below it, and
```
Comment (fchollet) [inline keras_core/ops/numpy.py:4859]: Use backticks

File: keras_core/ops/numpy.py [hunk 24]
```diff
@@ -3732,6 +4881,19 @@ def compute_output_spec(self, x):
...
 def tril(x, k=0):
+    """Return lower triangle of a tensor.
+
+    For tensors with `ndim` exceeding 2, `tril` will apply to the
+    final two axes.
+
+    Args:
+        x: Input tensor.
+        k: Diagonal above which to zero elements. Defaults to 0, the
+            main diagonal. k < 0 is below it, and k > 0 is above it.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:4892]: Use backticks

File: keras_core/ops/numpy.py [hunk 25]
```diff
@@ -3751,6 +4913,19 @@ def compute_output_spec(self, x):
...
 def triu(x, k=0):
+    """Return upper triangle of a tensor.
+
+    For tensors with `ndim` exceeding 2, `triu` will apply to the
+    final two axes.
+
+    Args:
+        x: Input tensor.
+        k: Diagonal below which to zero elements. Defaults to 0, the
+            main diagonal. k < 0 is below it, and k > 0 is above it.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:4924]: Here too

File: keras_core/ops/numpy.py [hunk 26]
```diff
@@ -3817,6 +5015,17 @@ def compute_output_spec(self, condition, x1, x2):
 
 @keras_core_export(["keras_core.ops.where", "keras_core.ops.numpy.where"])
 def where(condition, x1, x2):
+    """Return elements chosen from `x1` or `x2` depending on `condition`.
+
+    Args:
+        condition: Where True, yield `x1`, otherwise yield `x2`.
+        x1: Values from which to choose when `condition` is True.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:5022]: Backticks for True/False

File: keras_core/ops/numpy.py [hunk 27]
```diff
@@ -4097,6 +5413,18 @@ def compute_output_spec(self, x):
 
 @keras_core_export(["keras_core.ops.sum", "keras_core.ops.numpy.sum"])
 def sum(x, axis=None, keepdims=False):
+    """Sum of a tensor over the given axes.
+
+    Args:
+        x: Input tensor.
+        axis: Axis or axes along which the sum is computed. The default is to
+            compute the sum of the flattened tensor.
+        keepdims: If this is set to True, the axes which are reduced are left
```
Comment (fchollet) [inline keras_core/ops/numpy.py:5422]: Same here

File: keras_core/ops/numpy.py [hunk 28]
```diff
@@ -4140,6 +5486,19 @@ def compute_output_spec(self, N, M=None, k=0, dtype="float32"):
 
 @keras_core_export(["keras_core.ops.eye", "keras_core.ops.numpy.eye"])
 def eye(N, M=None, k=0, dtype="float32"):
+    """Return a 2-D tensor with ones on the diagonal and zeros elsewhere.
+
+    Args:
+        N: Number of rows in the output.
+        M: Number of columns in the output. If None, defaults to N.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:5493]: Backticks for None

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: keras_core/ops/numpy.py [hunk 29]
```diff
@@ -2788,6 +3522,19 @@ def compute_output_spec(self, x):
 
 @keras_core_export(["keras_core.ops.min", "keras_core.ops.numpy.min"])
 def min(x, axis=None, keepdims=False, initial=None):
+    """Return the minimum of a tensor or minimum along an axis.
+
+    Args:
+        x: Input tensor.
+        axis: Axis or axes along which to operate. By default, flattened input
+            is used.
+        keepdims: If this is set to True, the axes which are reduced are left
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3531]: Still missing one here :)

File: keras_core/ops/numpy.py [hunk 30]
```diff
@@ -2418,6 +2949,36 @@ def compute_output_spec(self, start, stop):
...
+        start: The starting value of the sequence.
+        stop: The end value of the sequence, unless `endpoint` is set to
+            False. In that case, the sequence consists of all but the last
+            of `num + 1` evenly spaced samples, so that `stop` is excluded.
+            Note that the step size changes when `endpoint` is `False`.
+        num: Number of samples to generate. Default is 50. Must be
+            non-negative.
+        endpoint: If `True`, `stop` is the last sample. Otherwise, it is
+            not included. Default is `True`.
+        retstep: If `True`, return (`samples`, `step`), where `step` is the
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2969]: Use backticks around the entire tuple, not the elements

File: keras_core/ops/numpy.py [hunk 31]
```diff
@@ -2622,6 +3261,31 @@ def compute_output_spec(self, start, stop):
...
+    `base ** stop` (see `endpoint` below).
+
+    Args:
+        start: The starting value of the sequence.
+        stop: The final value of the sequence, unless `endpoint` is `False`.
+            In that case, `num + 1` values are spaced over the interval in
+            log-space, of which all but the last (a sequence of length `num`)
+            are returned.
+        num: Number of samples to generate. Default is 50.
+        endpoint: If True, `stop` is the last sample. Otherwise, it is not
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3276]: Here too

File: keras_core/ops/numpy.py [hunk 32]
```diff
@@ -2529,6 +3136,17 @@ def compute_output_spec(self, x1, x2):
     ]
 )
 def logical_and(x1, x2):
+    """Computes the element-wise logical AND of the given input tensors.
+
+    Zeros are treated as False and non-zeros are treated as `True`.
```
Comment (fchollet) [inline keras_core/ops/numpy.py:3141]: Use backticks for False

File: keras_core/ops/numpy.py [hunk 33]
```diff
@@ -2418,6 +2949,36 @@ def compute_output_spec(self, start, stop):
...
+
+    Returns `num` evenly spaced samples, calculated over the interval
+    `[start, stop]`.
+
+    The endpoint of the interval can optionally be excluded.
+
+    Args:
+        start: The starting value of the sequence.
+        stop: The end value of the sequence, unless `endpoint` is set to
+            False. In that case, the sequence consists of all but the last
```
Comment (fchollet) [inline keras_core/ops/numpy.py:2962]: Here too

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

**Comment by FayazRahman** [comment]
Oops, must've got them all now

**Review by fchollet — APPROVED** [review]
Nice, thanks for the contribution!

**Pull request merged**